File: workflow/scripts/prune_tree.py
import argparse
import os
import pandas as pd
import numpy as np
from Bio import AlignIO
from dendropy import Tree

def main( args ):
    # Load metadata and generate strain - gisaid id dictionary.
    md = pd.read_csv( args.metadata, sep="\t" )
    print( "Entries in metadata: {}".format( len( md ) ) )
    md_gisaid_list = md["gisaid_epi_isl"].to_list()
    gisaid_dict = md.loc[~md["gisaid_epi_isl"].isna(), ["strain", "gisaid_epi_isl"]]
    gisaid_dict = gisaid_dict.set_index( "gisaid_epi_isl" )
    gisaid_dict = gisaid_dict["strain"].to_dict()

    # Load alignment
    alignment = AlignIO.read( args.alignment, "fasta" )
    print( "Sequences in alignment: {}".format( len( alignment ) ) )
    alignment_list = [i.name for i in alignment]

    # Load tree
    tree = Tree.get( path=args.tree, schema="newick" )
    print( "Leaves in tree: {}".format( len( tree.taxon_namespace ) ) )

    # Determine leaves which names cannot be assigned.
    tree_leaves = [i.label for i in tree.taxon_namespace]
    tree_leaves = [i.replace( " ", "_" ) for i in tree_leaves]
    leaf_missing_md = np.setdiff1d( tree_leaves, md_gisaid_list )

    # Remove leaves identified
    print( "Leaves in tree but not in metadata: {}".format( len( leaf_missing_md ) ) )
    tree = tree.extract_tree_without_taxa_labels( [i.replace( "_", " " ) for i in leaf_missing_md] )
    tree.purge_taxon_namespace()
    print( "Leaves in tree after pruning: {}".format( len( tree.taxon_namespace ) ) )

    # Rename leaves to match metadata and alignment
    print( "Renaming leaves to match metadata and alignment... ", end="" )
    leaves = list()
    for i in tree.taxon_namespace:
        i.label = gisaid_dict[i.label.replace( " ", "_" )]
        leaves.append( i.label )
    print( "Done" )

    # Remove leaves that aren't in alignment
    leaf_missing_align = np.setdiff1d( leaves, alignment_list )
    print( "Leaves in tree but not in alignment: {}".format( len( leaf_missing_align ) ) )
    tree = tree.extract_tree_without_taxa_labels( leaf_missing_align )
    tree.purge_taxon_namespace()
    print( "Leaves in tree after pruning: {}".format( len( tree.taxon_namespace ) ) )

    # Update tree_leaves list
    tree_leaves = [i.label for i in tree.taxon_namespace]

    # Filter alignment to tips in tree
    tree_alignment = list()
    for i in alignment:
        if i.name in tree_leaves:
            tree_alignment.append( i )
    tree_alignment = AlignIO.MultipleSeqAlignment( tree_alignment )

    # Filter metadata to tips in tree
    tree_md = md.loc[md["strain"].isin( [i.name for i in tree_alignment] )]

    # Filter metadata and alignment to query
    query_md = md.loc[md["interest"] == "interest"]
    interests = query_md["strain"].to_list()
    query_alignment = [i for i in alignment if i.name in interests]
    query_alignment = AlignIO.MultipleSeqAlignment( query_alignment )

    # Write files to disk
    tree.write( path=os.path.join( args.outdir, "global.tree" ), schema="newick" )
    AlignIO.write( tree_alignment, os.path.join( args.outdir, "alignment.fasta" ), "fasta" )
    tree_md.to_csv( os.path.join( args.outdir, "metadata.csv" ), index=False )
    AlignIO.write( query_alignment, os.path.join( args.outdir, "query.fasta" ), "fasta" )
    query_md.to_csv( os.path.join( args.outdir, "query.csv" ), index=False )

if __name__ == "__main__":
    parser = argparse.ArgumentParser( description="Prunes input tree to match alignment, metadata, and focus" )

    # Initialize optional arguments
    parser.add_argument( "-o", "--outdir", help="ouput directory" )
    parser.add_argument( "-m", "--metadata", help="input metdata containing at least name, gisaid_id, and interest field" )
    parser.add_argument( "-t", "--tree", help="input tree containing GISAID ids as tip names" )
    parser.add_argument( "-a", "--alignment", help="input filtered and masked alignment" )

    arguments = parser.parse_args()

    main( arguments )

File: workflow/scripts/prune_tree.R
library( ape )
library( Biostrings )

args = commandArgs( trailingOnly=TRUE )
output = args[4]

# Load data
tree <- read.tree( args[1] )
aln <- readDNAMultipleAlignment( args[2], format="fasta" )
md <- read.csv( args[3], sep="\t" )

# Rename tips in tree and remove tips which don't have an entry in tree
tree$tip.label[which(tree$tip.label %in% md$gisaid_epi_isl)] <- as.character( md$strain[match(tree$tip.label[which(tree$tip.label %in% md$gisaid_epi_isl)], md$gisaid_epi_isl)] )
print( cat( length( tree$tip.label[startsWith( tree$tip.label, "EPI")] ), "tips removed because of lack of metadata entry\n" ) )
tree <- drop.tip( tree, tree$tip.label[startsWith( tree$tip.label, "EPI")])

# Remove leaves that aren't in alignment. Output global tree
print( cat( length( setdiff( tree$tip.label, rownames( aln ) ) ), "tips removed because of lack of entry in alignment\n" ) )
tree <- drop.tip( tree, setdiff( tree$tip.label, rownames( aln ) ) )
write.tree( tree, file=file.path( output, "global.tree" ) )

print( cat( length( tree$tip.label ), "tips remaining in tree\n" ) )

# Write tips in tree to file
write( tree$tip.label, file.path( output, "alignment.txt" ) )
#rowmask( aln, invert=TRUE ) <- which( rownames( aln ) %in% tree$tip.label)
#writeXStringSet( x=as( aln, "XStringSet"), filepath=file.path( output, "alignment.fasta" ), format="fasta" )
#rowmask( aln ) <- NULL

# Filter MD to tips in tree.output and write to file
write.csv( subset( md, md$strain %in% tree$tip.label), file=file.path( output, "metadata.csv" ), row.names=FALSE )

# Filter metadata and alignment to query and write to file.
query_md <- subset( md, interest=="interest" )
write.csv( query_md, file=file.path( output, "query.csv" ), row.names=FALSE )
#rowmask( aln, invert=TRUE ) <- which( rownames( aln ) %in% query_md$strain )
#writeXStringSet( x=as( aln, "XStringSet"), filepath=file.path( output, "query.fasta" ), format="fasta" )
write.table( query_md$strain, file.path( output, "query.txt" ), row.names=FALSE, col.names=FALSE, quote=FALSE )
print( cat( nrow( query_md ), "sequences added to query.\n" ) )
